vanetza/security/default_certificate_validator.cpp

```diff
@@ -0,0 +1,122 @@
+            StartAndEndValidity start_and_end = boost::get<StartAndEndValidity>(validity_restriction);
+
+            // check if certificate validity restriction timestamps are logically correct
+            if (start_and_end.start_validity >= start_and_end.end_validity) {
+                return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+            }
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            if (now < start_and_end.start_validity || now > start_and_end.end_validity) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_End) {
+            EndValidity end = boost::get<EndValidity>(validity_restriction);
+
+            // check if certificate is outdated
+            auto now = convert_time32(time_now);
+            if (now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_Start_And_Duration) {
+            StartAndDurationValidity start_and_duration = boost::get<StartAndDurationValidity>(validity_restriction);
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            std::chrono::seconds duration = start_and_duration.duration.to_seconds();
+            auto end = start_and_duration.start_validity + duration.count();
+            if (now < start_and_duration.start_validity || now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        }
+    }
+
+    // if not exactly one time constraint is given, we fail instead of considering it valid
+    if (1 != certificate_time_constraints) {
+        return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+    }
+
+    // check if subject_name is empty
+    if (0 != certificate.subject_info.subject_name.size()) {
+        return CertificateInvalidReason::INVALID_NAME;
+    }
+
+    // check signer info
+    if (get_type(certificate.signer_info) != SignerInfoType::Certificate_Digest_With_SHA256) {
+        return CertificateInvalidReason::INVALID_ROOT_HASH;
+    }
+
+    HashedId8 signer_hash = boost::get<HashedId8>(certificate.signer_info);
+    std::vector<Certificate> possible_signers = trust_store.find_by_id(signer_hash);
+
+    // try to extract ECDSA signature
+    boost::optional<EcdsaSignature> sig = extract_ecdsa_signature(certificate.signature);
+    if (!sig) {
+        return CertificateInvalidReason::MISSING_SIGNATURE;
+    }
+
+    // create buffer of certificate
+    ByteBuffer cert = convert_for_signing(certificate);
+
+    bool valid_signature = false;
+
+    for (Certificate& possible_signer : possible_signers) {
```

Comment:
if you do that, this iteration would always happen exactly once (i.e., you'd have exactly one verification call to the crypto backend).

Reply:
If there's no collision, right. If there's no certificate found, it will just fail with `UNKNOWN_SIGNER`.